New CSS optimization part 1

Nilvus opened the pull request:
I've thought about ways to optimize basics CSS from @AlicVB, so like in first commit (a wrong split in commit let you see some lines removed in first commit, useful, put back in second one! Sorry for that mess in readability here). At the same time, @dterrahe suggested in https://github.com/darktable-org/darktable/pull/8278#issuecomment-782603288 that CSS needs optimization. I had done that one year ago. Anyway, many things have been added since and @dterrahe is right: this is needed again. This PR is all about that.

**Main goals**

- Reduce as much as possible CSS file (to help maintenance and updates). I already see many lines not needed (for example, some border, padding and margin set to 0 that they already inherit from exact same settings from lines above)
- Be sure as much as possible, there will be no more (or less as possible) CSS tweaks needed for new added features in the UI
- Try as much as possible to have same things in the UI using same settings (it's already the case for main parts like most of actual buttons, entry, textview, checkboxes, etc. ; but it's not always the case). Quite a hard part!
- Make the UI scalable as much as possible (as suggested by @elstoc)
- Remove some no more used lines

This PR is WIP by now as a lot of work remains to do. I already update some things so If some of you could test and post feedback here. I could have break something.

And fix #8353 

**Todo-list (as a reminder):**

- ~~make all things scalable~~
- ~~improve quick access panel: reduce vertical space when using small fonts and be sure it remains good with taller fonts~~
- ~~find a way to have an hover effect in color labels icons in footer toolbar~~
- ~~try to find a better way to display combobox when they are not opened (maybe!)~~
- ~~some spacing in color picker module~~
- ~~make blend GUI more compact with CSS (part of @ralfbrown idea in #8350)~~
- ~~check section_label and spacing (think about first one)~~
- ~~as discussed in #8437: check min-width of comboboxes and spinboxes and test if this could be improved (check too if 3D lut root folder one in processing tab could be adjusted consistently with CSS)~~

dterrahe commented:
So far so good. I don't see obvious breakage when using large font sizes. With small fonts, there is still too much vertical space between widgets in iop modules, but I know you've only just started :-) Many thanks (partly in advance) for this big effort!

You've mentioned before that sometimes you have to work around/correct for layouting problems that are introduced in the code. If you need any help fixing stuff on that side, please ping me.

Nilvus commented:
All UI is now scalable. @elstoc: I know you want that, I'd like you test that, even if that PR is not over, scalable part is over. No more px is CSS!

And already more than 140 lines removed and a little improvement in color picker module. More things to come.

dterrahe commented on data/themes/darktable.css:
```diff
@@ -1312,42 +1166,37 @@ dialog scrollbar
 }
 
-#preferences_notebook scrolledwindow #pref_section  /* top margin in sub-category title in main content */
+#preferences_notebook scrolledwindow #pref_section :not(:first-child) /* top margin in sub-category title in main content */
```
Would it be possible to do something similar for `section_label` so that it automatically look good even at the top of a box without requiring this line in the code?
`gtk_style_context_add_class(context, "section_label_top");`

elstoc replied:
This is only for when the section label is at the top of an iop rather than when it is the first section label _in_ the iop (which might be after some other controls). But I suspect it could be achieved with pure CSS.

dterrahe replied:
When the section label is the first _child_ within a box. Could be that the box is on a notebook page, doesn't matter; if the first thing within that box is a section label, it should get special treatment.
But if there first is something else (slider, whatever) and _then_ a section label; even though it might be the first _section label_ it should just be treated the same as any following section labels.

:first-child sounds like it should do the right thing...

elstoc replied:
could be, yes

Nilvus replied:
first-child is exactly for that: setting a first child of a widget differently than following ones. Here, with not, I just apply to all except first child. About @section label, that should come. Many things remain to do.

elstoc commented on data/themes/darktable.css:
```diff
@@ -253,44 +253,36 @@ box *
 #button-canvas
 {
-  margin: 5px; /* not real pixels, this is used as a percent extra space */
+  margin: 0.35em; /* not real pixels, this is used as a percent extra space */
```
based on the comment I think perhaps this one needs to be left alone doesn't it?

Nilvus replied:
Gtk apply pixels from em depending on font size, so should not be a problem. On what I test, it's all good. I see no issue. I've tested with 11px font size and your small font size, 8.4px. But feel free to test too, I could have missed something.

elstoc replied:
My recollection was that somewhere (in the overlays maybe) the number of pixels was converted to a percentage in gtk. It may be that code has been removed now - perhaps @AlicVB can recall, but if that comment is not applicable any longer it should be removed.

Nilvus replied:
As I know, it's always pixels. But depending on how it sets, that do something like a percentage. Just check with Gtk Inspector Tool, and with margin set to 5px, you will see 5px. With 0.35em with 11px, you will see 5.0xxx px. But on those settings, that is considered to some percentage by adjusting space remaining depending on margin set, not real percentage. I don't know if I'm clear here. So, having 5px or 0.35em doesn't change what that is for. The only change is that we have the logic (not real pixels and used as a "percent" extra space) but scalable.

AlicVB replied:
I've just tested, and yes, the order seems : 
- margin is defined in css with em
- when retrieve by gtk functions, we get the corresponding pixels (internally computed by gtk)
- then we use that value as a percentage of the thumb size

So I don't any problem here (it's more an enhancement) apart from the comment which is not accurate anymore. Maybe something like `/* not real value, corresponding pixels size is used as a percent extra space */` but I'm sure there's better English speakers than me around ;)

elstoc replied:
Thanks @AlicVB. Is that the same for everywhere this comment appears in darktable.css? Comment should perhaps be something along the lines of `/* this em value is converted to pixels by Gtk and then used as a percentage of extra space around the element */`.

Nilvus replied:
Thanks for the suggestion @elstoc. Seems indeed more appropriate. Will update that soon.

Nilvus replied:
comment updated with your proposal @elstoc.

Nilvus commented:
Last commit improve spacing in quick access panel and all section label, including a more compact blend gui, following #8359 work from @dterrahe. It only remains a little issue in drawn mask line below, due to combobox added in the line (I've try to reduce that a little with CSS but not succeed to do more (no more padding/margin to work on):

![2021-03-06_23-10](https://user-images.githubusercontent.com/45535283/110222255-c6df1600-7ec8-11eb-88d1-517c4396832f.png)

elstoc commented:
I've done a little bit of testing this evening and some of the scaling looks a little too small with my standard 8.4pt font. The following is an example from current master followed by this PR

![Master](https://user-images.githubusercontent.com/9555491/110222336-61d7f000-7ec9-11eb-8ba6-987e9c04d126.png) ![PR](https://user-images.githubusercontent.com/9555491/110222340-6c928500-7ec9-11eb-96e7-b17ce363abcd.png)

I think here the interface is a little too compressed. Most of the icons in the parametric mask section are very small and the spacing between the controls is a little too tight even for me. This is especially noticeable at the bottom of the blend section where the distance between the final slider and the bottom of the module is very small. I wonder perhaps if we didn't go a bit far with the idea of "no pixel values" and maybe some of the UI really only makes sense at a relatively fixed size (icons mostly).

Also, at smaller font sizes I notice that the hover effect on the buttons in the quick access panel is rather one sided again:

![QA](https://user-images.githubusercontent.com/9555491/110222576-263e2580-7ecb-11eb-8230-746258293b38.png)

dterrahe commented:
> including a more compact blend gui, following #8359 work from @dterrahe. It only remains a little issue in drawn mask line below, due to combobox added in the line (I've try to reduce that a little with CSS but not succeed to do more (no more padding/margin to work on)

I tried to reduce this a little with#8425. I'm now seeing this:
![image](https://user-images.githubusercontent.com/1549490/110240214-c4210700-7f42-11eb-8e4f-17c2f4a69146.png)

GtkInspector tells me there's a 3.733px padding around the section_label boxes. Is that a problem on my side? It doesn't just take up vertical space, but also indents the labels (on the left) and the buttons (on the right) so that they don't align with the buttons directly below

Nilvus commented:
> I think here the interface is a little too compressed. Most of the icons in the parametric mask section are very small and the spacing between the controls is a little too tight even for me. This is especially noticeable at the bottom of the blend section where the distance between the final slider and the bottom of the module is very small. I wonder perhaps if we didn't go a bit far with the idea of "no pixel values" and maybe some of the UI really only makes sense at a relatively fixed size (icons mostly).
> 
> Also, at smaller font sizes I notice that the hover effect on the buttons in the quick access panel is rather one sided again:

Yes, you're right. Some lines need to not be scalable and remains at constant px. Will revert that. Thanks for testing!

Nilvus commented:
@elstoc: could you check again and confirm if last commit fix issues you had with that PR?

Nilvus commented:
> GtkInspector tells me there's a 3.733px padding around the section_label boxes. Is that a problem on my side? It doesn't just take up vertical space, but also indents the labels (on the left) and the buttons (on the right) so that they don't align with the buttons directly below

It's set with CSS. Just tested your PR and it's better. Thanks. Will check and test that.

Nilvus commented:
> GtkInspector tells me there's a 3.733px padding around the section_label boxes. Is that a problem on my side? It doesn't just take up vertical space, but also indents the labels (on the left) and the buttons (on the right) so that they don't align with the buttons directly below

@dterrahe: adjustments made based on your PR (so last master now) and indent are fixed. This will be in next following commit.

Nilvus commented:
Last commit fix blending padding, add hover effect in color labels on footer-headerbar in lighttable view and change combobox to make them a little more visible. See:

![2021-03-07_22-01](https://user-images.githubusercontent.com/45535283/110254757-64e5e580-7f88-11eb-8119-e524786ef756.png)

![2021-03-07_22-01_1](https://user-images.githubusercontent.com/45535283/110254763-68796c80-7f88-11eb-8bd0-a9e7d8879c9d.png)

I hope you will like them.

elstoc commented:
The icon sizes are definitely better now.  Thanks.

Mark-64 commented:
I quickly tested the PR in Windows 10 / HiDPI and I found no obvious flaws.
It look more compact and overall nicer to me.
Good job!

Nilvus commented:
As I have quite less time actually and probably for some time, I propose this PR to review and merge if all is ok. Will remains some parts to optimize, like: 

- optimize if possible CSS code and UI in:
  - modulegroups
  - lighttable
  - new import module

And probably a little more. Will do that when I will have time...

@TurboGit: you could review and merge. More feedback by merging it will be needed, especially to check I don't break anything.

elstoc commented:
Looks a bit more even now thanks, though I note that when not expanded, the blending bar is no longer flush with the bottom of the module:

![Screenshot_2021-03-18_21-36-54](https://user-images.githubusercontent.com/9555491/111701001-13fca980-8832-11eb-8b95-99a237ef777b.png)

One other thing I noticed though is that when I changed my font size, the border around thumbnail images in the lighttable changed as well. It doesn't seem right to scale that border with font size - that was the 'pixels that aren't real pixels' setting we discussed earlier.

Nilvus commented:
Thanks for testing @elstoc.

> I think the modules still need a little more space around the edges. You can see in the following screenshot that there's not a lot of space between the left edge of the module and the start of the sliders (contrasting with the larger space on the right). Similarly between the black level correction slider and the blend mode selection, and between the mask contrast slider and the bottom of the module.

Indeed you're right. Should be better with last commit.

> Similar thing on the quick access panel - everything just seems that little bit too close together. Perhaps there's something to be said for some "pixel" padding on these as well (though it may be fine with just a bit of extra space in ems).

I have to admit that I do not test so much that part. Should be better too with last commit. And all that allow me to optimize a little more CSS code by removing thirteen more lines.
 
> I'm not incredibly fond of the gradient on the comboboxes in the preferences window either - I think a plain dark background is fine - though that's purely personal preference.

I found them better than it is actually in master and dark background would not suit all places. By now, it's the better compromise I found to have same background in all comboboxes (even if background around combobox is not the same. And that's the hard deal here!).

elstoc commented:
> I found them better than it is actually in master and dark background would not suit all places.

Doesn't this just impact the preferences dialog? Or are there more elsewhere that I'm missing?

Nilvus commented:
> Looks a bit more even now thanks, though I note that when not expanded, the blending bar is no longer flush with the bottom of the module:

Oh shit, forget to check that. So I see no possibility to have bottom margin or padding when expanded (except if a Gtk tweak to apply it only when expanded is possible). And I tested all things I know, even CSS selector like last-child, last-of-type or nth-last-child to apply only on last bauhaus-slider. Even that let this glitch. So I remove it.
 
> One other thing I noticed though is that when I changed my font size, the border around thumbnail images in the lighttable changed as well. It doesn't seem right to scale that border with font size - that was the 'pixels that aren't real pixels' setting we discussed earlier.

I had reverted all "real pixels" quite some days ago. Such border was not on those. I've checked with your 8.4px font size with grey theme and don't see any issue. Or maybe I don't understand correctly. It's good for me.

Nilvus commented:
> Doesn't this just impact the preferences dialog? Or are there more elsewhere that I'm missing? I'm just talking about them having the same background as the other widgets in preferences.

Comboboxes in header toolbar (to filter images). Having same background in preferences was my goal initially but as I try to optimize and avoid having 2 or more differents settings for same field, I have found that this "gradient" was best compromise to show comboboxes in all part of darktable window (where they are actually at least) and so made a gradient near background of all other widgets in preferences AND having combobox in header toolbar good seeing. Not easy.

elstoc commented:
Ok. I see the desire to make the CSS generic but there will be circumstances when it's not possible. IMO two separate combobox style definitions is the lesser evil, but let's see what others think.

Nilvus commented:
> Ok. I see the desire to make the CSS generic but there will be circumstances when it's not possible. IMO two separate combobox style definitions is the lesser evil, but let's see what others think.

Of course, sometime it's not possible. But I think my gradient combobox is not so bad. The better will be to test and if no more feedback here, the master merge will be the hard test. If some people complains, I will do different setting for those comboboxes.

elstoc commented:
> Oh shit, forget to check that. So I see no possibility to have bottom margin or padding when expanded (except if a Gtk tweak to apply it only when expanded is possible). And I tested all things I know, even CSS selector like last-child, last-of-type or nth-last-child to apply only on last bauhaus-slider. Even that let this glitch. So I remove it in following commit update.

The following works for me (though the values are purely for testing purposes)
```
#blending-box
{  margin-bottom: 0.75em;  }

#iop-plugin-ui-main
{  margin-bottom: 0.75em;  }
```

> I've checked with your 8.4px font size with grey theme and don't see any issue. Or maybe I don't understand correctly. It's good for me.

Go into the lighttable and open up preferences. Change the font size to 8.4. Check the border between thumbnail images. Change the font size to 12 and exit. The border between thumbnail images changes.

8.4
![Screenshot_2021-03-18_23-08-45](https://user-images.githubusercontent.com/9555491/111709121-f97cfd00-883e-11eb-9998-b0a141910dcb.png)

12
![Screenshot_2021-03-18_23-09-08](https://user-images.githubusercontent.com/9555491/111709130-feda4780-883e-11eb-88f4-06ac1f08764e.png)

Nilvus commented:
> The following works for me (though the values are purely for testing purposes)
> 
> ```
> #blending-box
> {  margin-bottom: 0.75em;  }
> 
> #iop-plugin-ui-main
> {  margin-bottom: 0.75em;  }
> ```

That add bottom margin in all blending-box and not only to last one. I've of course begin with that.
Anyway, I found it good as it is even if a little more padding, only in last blending-box would be better.

> > go into the lighttable and open up preferences. Change the font size to 8.4. Check the border between thumbnail images. Change the font size to 12 and exit. The border between thumbnail images changes.

So that was I understood. Such border is scalable and I don't see the problem here. It's good for me. This fit with the rest of the scalable UI.

elstoc commented:
> This fit with the rest of the scalable UI

Maybe. People can use DPI to scale the whole UI though. It's a minor thing but it feels wrong to scale image stuff with font size.

Nilvus commented:
> > This fit with the rest of the scalable UI
> 
> Maybe. People can use DPI to scale the whole UI though. It's a minor thing but it feels wrong to scale image stuff with font size. It makes more sense with the rest of the UI because there's text everywhere on it. Again a minor thing that I can tweak in prefs if needed.

That's not really image stuff but I understand what is wrong for you. So I revert back (last commit) thumb borders to px.

> I see what you mean - last-of-type would be the one to use but it doesn't work in Gtk. I do think that looks a bit nicer with a bit of space between each of the blending sections, perhaps not with 0.75em - 0.5em looks good to me.

You're right, it's better with some bottom margin in blending box (I set it to 0.28em). See last commit.

Thanks for all your feedback @elstoc. As always, it's helpful.

@TurboGit: if it's ok for @elstoc, this PR is ready to be merged.

elstoc commented:
Yep looks good to me. Thanks @Nilvus.

TurboGit reviewed (approved):
Thanks!
